ocaml/xapi/console.ml

```diff
@@ -126,7 +127,309 @@ let address_of_console __context console : address option =
+      ; Unixext.poll_event_timeout_ms
+      }
+    in
+    log_in_file
+      (Printf.sprintf
+         "[get_init_console_state] initial_state: is_idle_timeout_set=%b, \
+          is_idle=%b, poll_event_timeout_ms=%d"
+         initial_state.is_idle_timeout_set initial_state.is_idle
+         initial_state.poll_event_timeout_ms
+      ) ;
+    initial_state
+
+  (* Creates a VNC console idle timeout callback closure to determine is_idle *)
+  let create_vnc_console_idle_timeout_callback ~__context ~vm
+      initial_console_state =
+    let simple_rfb_parser = RFB_msg_type_parser.create () in
+    let current_console_state = ref initial_console_state in
+    let idle_timer = ref None in
+
+    (* Helper function to start/restart the idle timer *)
+    let start_idle_timer reason =
+      let _, _, idle_timeout_seconds = read_console_config ~__context ~vm in
+      let timeout_duration =
+        Timer.s_to_span (Int64.to_float idle_timeout_seconds) |> Option.get
+      in
+      idle_timer := Some (Timer.start ~duration:timeout_duration) ;
+      log_in_file
+        (Printf.sprintf "[idle_timeout] Starting/ restarting idle timer %s"
+           reason
+        )
+    in
+
+    (* Start the timer immediately if idle timeout is enabled *)
+    if initial_console_state.Unixext.is_idle_timeout_set then
+      start_idle_timer "on callback creation" ;
+    fun received_data_opt ->
+      (* First update any configuration that might have changed *)
+      let is_idle_timeout_set, poll_event_timeout_ms, _ =
+        read_console_config ~__context ~vm
+      in
+      current_console_state :=
+        {
+          !current_console_state with
+          Unixext.is_idle_timeout_set
+        ; Unixext.poll_event_timeout_ms
+        } ;
+
+      if !current_console_state.Unixext.is_idle_timeout_set then
+        (* Though the callback is only invoked when the timeout is set,
+           we still check here in case the setting changed from true to false.
+           The case from false to true would never reach here *)
+        match !idle_timer with
+        (* Check if idle timer has expired first, regardless of received data *)
+        | Some timer when Timer.has_expired timer ->
+            log_in_file
+              "[idle_timeout] Idle timeout reached - session should be \
+               terminated" ;
+            {!current_console_state with Unixext.is_idle= true}
+        | _ -> (
+          (* Timer not expired or not started yet, continue processing *)
+          match received_data_opt with
+          | None ->
+              !current_console_state
+          | Some data ->
+              RFB_msg_type_parser.process_data simple_rfb_parser data 0 ;
+
+              (* Check if we received KeyEvent (4) or PointerEvent (5) messages *)
+              let has_activity =
+                List.exists
+                  (fun msg_type -> msg_type = 4 || msg_type = 5)
+                  simple_rfb_parser.received_msg_types
```

Comment:
I think the logic could be as simple (two lines maybe) as:
1. refresh the state with a new`Mtime_clock.counter ()` when parsed a `KeyEvent` or `PointerEven` message.
2. after all parsing in one callback call, check the elapsed time with `Mtime_clock.count` against the idle timeout config.